cmd/opentelemetry-prometheus-sidecar/main.go

```diff
@@ -237,52 +296,28 @@ func Main() bool {
 	level.Debug(logger).Log("msg", "entering run state")
 	healthChecker.SetRunning()
 
-	// Run two inter-depdendent components:
-	// (1) Prometheus reader
-	// (2) Queue manager
-	// TODO: Replace this with x/sync/errgroup
-	var g run.Group
-	{
-		g.Add(
-			func() error {
-				level.Info(logger).Log("msg", "starting Prometheus reader")
-				err = prometheusReader.Run(ctx, startOffset)
-				return err
-			},
-			func(err error) {
-				// Prometheus reader needs to be stopped before closing the TSDB
-				// so that it doesn't try to write samples to a closed storage.
-				// See the use of `stopCh` below to explain how this works.
-				level.Info(logger).Log("msg", "stopping Prometheus reader")
-				cancelMain()
-			},
-		)
-	}
-	{
-		stopCh := make(chan struct{})
-		g.Add(
-			func() error {
-				if err := queueManager.Start(); err != nil {
-					return err
+	g := getRunGroup(ctx, cancelMain, logger, startOffset, prometheusReader, queueManager)
+
+	for {
+		if err := g.Run(); err != nil {
+			// TODO: don't use a string compare here
+			if strings.Contains(err.Error(), "truncated WAL segment") {
+				startOffset, err = readWriteStartOffset(cfg, logger)
```

Comment:
I was wondering if we should somehow advance the WAL segment number returned by `readWriteStartOffset()` after one of these corruptions, otherwise won't we end up in the same place?

Reply:
so what i see happening, is the `offset` from the tailer is less than the `startOffset` when the reader is restarted,  causing it to skip until the next boundary in this code here, which appears to have the desired effect
https://github.com/lightstep/opentelemetry-prometheus-sidecar/blob/719e94051e0d629490d9aee6d9ea9d4d524dc7c2/retrieval/manager.go#L169-L177

not super familiar with this codepath though so it's quite possible i'm missing something

Reply:
(Discussed this offline with @codeboten:)

We think the sequence of events is:

1. Prom and Sidecar both start. Sidecar enters a wait state on a partial page, sits there asking Prom what page is next.
2. Prom begins its recovery
3. Prom erases the file we're at EOF on
4. Prom sets the current segment to N+k for k>1
5. Sidecar restarts
6. Sidecar sees N+k as the oldest log, then skips forward in retrieval/manager.go

This is :100:!